go-controller/pkg/ovn/dns-name-resolver/egressfirewall_external_dns.go

```diff
@@ -0,0 +1,386 @@
+	}
+	extEgDNS.dnsNameResolverController = controller.NewController[ocpnetworkapiv1alpha1.DNSNameResolver]("dnc-dns-controller", dnsConfig)
+}
+
+// dnsNeedsUpdate returns true if a dns name resolver object is either added
+// or deleted. If a dns name resolver is updated, then dnsNeedsUpdate returns
+// true if the .status of the object is modified.
+func dnsNeedsUpdate(oldObj, newObj *ocpnetworkapiv1alpha1.DNSNameResolver) bool {
+	if oldObj == nil || newObj == nil {
+		return true
+	}
+	return !reflect.DeepEqual(oldObj.Status, newObj.Status)
+}
+
+// reconcileDNSNameResolver reconciles a DNSNameResolver object. If an object was deleted,
+// then egressFirewallExternalDNS.Delete is called for the corresponding DNS name. If an
+// object is added/updated, then egressFirewallExternalDNS.Add is called for the
+// corresponding DNS name along with the associated addresses.
+func (extEgDNS *ExternalEgressDNS) reconcileDNSNameResolver(key string) error {
+	// Split the key in namespace and name of the corresponding object.
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("reconcileDNSNameResolver failed to split meta namespace cache key %s for dns name resolver: %v", key, err)
+		return nil
+	}
+	// Fetch the dns name resolver object using the name and namespace.
+	obj, err := extEgDNS.dnsLister.DNSNameResolvers(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// DNSNameResolver object was deleted.
+
+			// Get the corresponding DNS name, if avialable.
+			dnsName, found := extEgDNS.resolverToDNSName[name]
+			// If the DNS name was not found then return.
+			if !found {
+				return nil
+			}
+
+			// Delete the address set associated with the DNS name.
+			var err error
+			waitErr := wait.PollUntilContextTimeout(context.Background(), 100*time.Millisecond, 1*time.Second, true, func(ctx context.Context) (bool, error) {
+				deleteResp := extEgDNS.Delete(DeleteRequest{DNSName: dnsName})
+				err = deleteResp.Err
+				if err != nil {
+					return false, nil
+				}
+				return true, nil
+			})
+			if waitErr != nil {
+				errs := []error{waitErr}
+				if err != nil {
+					errs = append(errs, err)
+				}
+				return fmt.Errorf("error with reconcileDNSNameResolver - %v", errors.NewAggregate(errs))
+			}
+
+			// Delete the corresponding entries from the objDNSName
+			// and the dnsNameObj maps.
+			delete(extEgDNS.resolverToDNSName, name)
+			delete(extEgDNS.dnsNameToResolver, dnsName)
+
+			return nil
+		}
+		return fmt.Errorf("reconcileDNSNameResolver failed to fetch dns name resolver %s in namespace %s", name, namespace)
+	}
+
+	// DNSNameResolver object was added/updated.
+
+	// Check if the object name corresponding to the DNS name exists
+	// or not. If it exists, whether the object name matches with the
+	// exisiting extry.
+	dnsName := string(obj.Spec.Name)
+	objName, exists := extEgDNS.dnsNameToResolver[dnsName]
+	if exists && objName != name {
```

Comment:
should this be `objName == name`? different names sound like an error?

Reply:
Existing object name not matching the current object name can happen if a user creates the DNSNameResolver object for the same DNS name. The cluster-manager will ensure that such objects are removed from the cluster, however, the node will still receive the events related to the unwarranted object, It is not an error on the node side and any event related to such objects should be ignored.